File: widget/src/main/java/soup/widget/CheckedImageView.java
package soup.widget;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.accessibility.AccessibilityEvent;
import android.view.accessibility.AccessibilityNodeInfo;
import android.widget.Checkable;
import android.widget.ImageView;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

@SuppressLint("AppCompatCustomView")
public class CheckedImageView extends ImageView implements Checkable {
    private boolean mChecked;
    private boolean mBroadcasting;

    private OnCheckedChangeListener mOnCheckedChangeListener;

    private static final int[] CHECKED_STATE_SET = {
            android.R.attr.state_checked
    };

    public CheckedImageView(Context context) {
        this(context, null);
    }

    public CheckedImageView(Context context, AttributeSet attrs) {
        this(context, attrs, R.attr.checkedImageViewStyle);
    }

    public CheckedImageView(Context context, AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    public CheckedImageView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        final TypedArray a = context.obtainStyledAttributes(
                attrs, R.styleable.CheckedImageView, defStyleAttr, defStyleRes);

        final boolean checked = a.getBoolean(R.styleable.CheckedImageView_checked, false);
        setChecked(checked);

        a.recycle();
    }

    @Override
    public void toggle() {
        setChecked(!mChecked);
    }

    @Override
    public boolean isChecked() {
        return mChecked;
    }

    @Override
    public void setChecked(boolean checked) {
        if (mChecked != checked) {
            mChecked = checked;
            refreshDrawableState();
            notifyViewAccessibilityStateChangedIfNeeded(
                    AccessibilityEvent.CONTENT_CHANGE_TYPE_UNDEFINED);

            // Avoid infinite recursions if setChecked() is called from a listener
            if (mBroadcasting) {
                return;
            }

            mBroadcasting = true;
            if (mOnCheckedChangeListener != null) {
                mOnCheckedChangeListener.onCheckedChanged(this, mChecked);
            }

            mBroadcasting = false;
        }
    }

    public void setOnCheckedChangeListener(OnCheckedChangeListener listener) {
        mOnCheckedChangeListener = listener;
    }

    public static interface OnCheckedChangeListener {
        void onCheckedChanged(CheckedImageView buttonView, boolean isChecked);
    }

    @Override
    public int[] onCreateDrawableState(int extraSpace) {
        final int[] drawableState = super.onCreateDrawableState(extraSpace + 1);
        if (isChecked()) {
            mergeDrawableStates(drawableState, CHECKED_STATE_SET);
        }
        return drawableState;
    }

    @Override
    public void onInitializeAccessibilityEvent(AccessibilityEvent event) {
        super.onInitializeAccessibilityEvent(event);
        event.setClassName(CheckedImageView.class.getName());
        event.setChecked(mChecked);
    }

    @Override
    public void onInitializeAccessibilityNodeInfo(AccessibilityNodeInfo info) {
        super.onInitializeAccessibilityNodeInfo(info);
        info.setClassName(CheckedImageView.class.getName());
        info.setCheckable(true);
        info.setChecked(mChecked);
    }

    /** call @hide API in View.java using reflection */
    private void notifyViewAccessibilityStateChangedIfNeeded(int changeType) {
        try {
            Method method = android.view.View.class.getMethod(
                    "notifyViewAccessibilityStateChangedIfNeeded",
                    new Class[]{int.class});
            method.invoke(this, new Object[]{changeType});
        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
            // nothing
        }
    }
}

File: widget/src/main/java/soup/widget/CheckedImageView.kt
package soup.widget

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.accessibility.AccessibilityEvent
import android.view.accessibility.AccessibilityNodeInfo
import android.widget.Checkable
import android.widget.ImageView

import java.lang.reflect.InvocationTargetException

@SuppressLint("AppCompatCustomView")
class CheckedImageView @JvmOverloads constructor(context: Context,
                                                 attrs: AttributeSet? = null,
                                                 defStyleAttr: Int = R.attr.checkedImageViewStyle,
                                                 defStyleRes: Int = 0) :
        ImageView(context, attrs, defStyleAttr, defStyleRes), Checkable {

    private var checked: Boolean = false
    private var broadcasting: Boolean = false

    private var onCheckedChangeListener: OnCheckedChangeListener? = null

    init {
        val a = context.obtainStyledAttributes(
                attrs, R.styleable.CheckedImageView, defStyleAttr, defStyleRes)

        val checked = a.getBoolean(R.styleable.CheckedImageView_checked, false)
        isChecked = checked

        a.recycle()
    }

    override fun toggle() {
        isChecked = !checked
    }

    override fun isChecked(): Boolean {
        return checked
    }

    override fun setChecked(checked: Boolean) {
        if (this.checked != checked) {
            this.checked = checked
            refreshDrawableState()
            notifyViewAccessibilityStateChangedIfNeeded(
                    AccessibilityEvent.CONTENT_CHANGE_TYPE_UNDEFINED)

            // Avoid infinite recursions if setChecked() is called from a listener
            if (broadcasting) {
                return
            }

            broadcasting = true
            onCheckedChangeListener?.onCheckedChanged(this, this.checked)
            broadcasting = false
        }
    }

    fun setOnCheckedChangeListener(listener: OnCheckedChangeListener) {
        onCheckedChangeListener = listener
    }

    interface OnCheckedChangeListener {
        fun onCheckedChanged(buttonView: CheckedImageView, isChecked: Boolean)
    }

    override fun onCreateDrawableState(extraSpace: Int): IntArray {
        return super.onCreateDrawableState(extraSpace + 1).also {
            if (isChecked) {
                View.mergeDrawableStates(it, CHECKED_STATE_SET)
            }
        }
    }

    override fun onInitializeAccessibilityEvent(event: AccessibilityEvent) {
        super.onInitializeAccessibilityEvent(event)
        with(event) {
            className = CheckedImageView::class.java.name
            isChecked = checked
        }
    }

    override fun onInitializeAccessibilityNodeInfo(info: AccessibilityNodeInfo) {
        super.onInitializeAccessibilityNodeInfo(info)
        with(info) {
            className = CheckedImageView::class.java.name
            isCheckable = true
            isChecked = checked
        }
    }

    /** call @hide API in View.java using reflection  */
    private fun notifyViewAccessibilityStateChangedIfNeeded(changeType: Int) {
        try {
            val method = android.view.View::class.java.getMethod(
                    "notifyViewAccessibilityStateChangedIfNeeded",
                    *arrayOf<Class<*>>(Int::class.javaPrimitiveType!!))
            method.invoke(this, *arrayOf<Any>(changeType))
        } catch (e: NoSuchMethodException) {
        } catch (e: IllegalAccessException) {
        } catch (e: InvocationTargetException) {
        }
    }

    companion object {

        private val CHECKED_STATE_SET = intArrayOf(android.R.attr.state_checked)
    }
}